### PR TITLE
GPT2 SDPA inference pattern-matching for Inductor-CPU

### PR DESCRIPTION
**Pull request by sanchitintel**
### Summary
With this PR, SDPA pattern of GPT2 is being mapped to `torch.nn.functional.scaled_dot_product_attention`.
While GPT2 supports both a causal mask & an attention mask, this PR considers the case of attention mask being absent.
TorchBench inference workload for GPT2 also doesn't use an attention-mask.

This pattern's replacement is being disabled for CUDA because [CUDA AOT Inductor](https://github.com/pytorch/pytorch/actions/runs/8319111885/job/22762567770) CI job's `GPT2ForSequenceClassification` accuracy test failed, although all other trunk CUDA Inductor CI checks had passed.
Created #122429 to track that particular issue.

### CPU performance data with TorchBench
|MODEL |BATCH SIZE | DTYPE | BEFORE: Speedup over eager-mode with the default Inductor implementation | AFTER: Speedup over eager mode with SDPA op mapped| Perf boost = (AFTER - BEFORE)/BEFORE * 100|
|--------------------------|-------------|---------|-----------------------------|--------------------------|------------|
|hf_GPT2| 1 | FP32 | 1.522x | 1.791x| 17.67%|
|hf_GPT2| 1 | BF16 (AMP) | 1.795x | 2.387x| 32.98%|
|hf_GPT2| 2 | FP32 |  1.313x |1.629x | 19.3%|
|hf_GPT2|2| BF16 (AMP) | 1.556x | 1.924x | 23.65%|
|hf_GPT2_large| 1 | FP32 | 1.380x |1.585x | 12.93%|
|hf_GPT2_large| 1 | BF16 (AMP) | 1.208x | 1.567x | 22.91%|
|hf_GPT2_large| 2 | FP32 | 1.188x | 1.490x | 25.42%|
|hf_GPT2_large|2| BF16 (AMP) | 0.991x | 1.575x | 58.93%|

Machine - Intel(R) Xeon(R) Platinum 8468H (Xeon 4th gen Sapphire Rapids)
48 physical cores were used. Intel OpenMP & libtcmalloc were preloaded.

Example command -
```
 OMP_NUM_THREADS=48 MKL_NUM_THREADS=48 numactl --membind=0 --cpunodebind=0 -C 0-47 python benchmarks/dynamo/torchbench.py --performance --inference --inductor --float32 -dcpu --only hf_GPT2_large --freezing --batch-size 1
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 1]
```diff
@@ -453,6 +453,49 @@ def _sfdp_replacement_17(query, key, value, attn_mask, inv_scale, dropout_p):
...
+                value
+            )
+        ),
+        key,
+        value,
+    )
+
+
+def _sfdp_replacement_18(
+    query, key, value, inv_scale, causal_mask_value, causal_mask, dropout_p
```
Comment (Valentine233) [inline torch/_inductor/fx_passes/fuse_attention.py:479]: For `causal_mask_value`, SDPA only accepts `-inf`. Is it better to use `causal_mask_value` as a constant rather than an input, because this may cause accuracy issue? Similar as https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/fuse_attention.py#L425.

File: torch/_inductor/fx_passes/serialized_patterns/central_index.py [hunk 2]
```diff
@@ -1,5 +1,3 @@
-# mypy: ignore-errors
```
Comment (Valentine233) [inline torch/_inductor/fx_passes/serialized_patterns/central_index.py:1]: I suppose this should be kept.

File: torch/_inductor/fx_passes/fuse_attention.py (hunk 1 above)
Comment (sanchitintel) [reply torch/_inductor/fx_passes/fuse_attention.py:479]: Thanks for the info! I'm not using the value of `causal_mask_value` as a parameter in `torch.nn.functional.scaled_dot_product_attention`, but it was necessary to match the pattern.

In GPT2, `causal_mask_value` is `torch.finfo(dtype).min`, but the output of the SDPA op should be similar to it being `-float("inf")`, so I'm using the latter in the replacement pattern, similarly as the reference you provided.

I'll apply the other changes.

Comment (Valentine233) [reply torch/_inductor/fx_passes/fuse_attention.py:479]: I mean that we could remove `causal_mask_value` in the parameter lists and do `causal_mask_value = torch.finfo(dtype).min` in the pattern function.

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 3]
```diff
@@ -704,6 +750,20 @@ def _get_sfdp_patterns():
                 d,
                 _sfdp_extra_check(aten.div.Tensor),
             ),
+            (
+                _sfdp_pattern_18,
+                _sfdp_replacement_18,
+                [g(), g(), g(), c(), c(), m_bool()],
+                d,
+                _sfdp_extra_check(disable_cuda=True),
```
[the remark this answers is not in the document]
Comment (sanchitintel) [reply torch/_inductor/fx_passes/fuse_attention.py:759]: Done, thanks!

File: torch/_inductor/fx_passes/fuse_attention.py (hunk 1 above)
Comment (sanchitintel) [reply torch/_inductor/fx_passes/fuse_attention.py:479]: Thanks, but IIRC, you had a comment in a PR about the pattern not matching when `torch.finfo(dtype).min` was used. I'll try it, and will push changes, if it'd work.

Comment (Valentine233) [reply torch/_inductor/fx_passes/fuse_attention.py:479]: Do you indicate this [issue](https://github.com/pytorch/pytorch/issues/117735), in which the problem is `torch.tensor()`? I suppose `torch.finfo(scores.dtype).min` should work and you could have a try. Thanks~

Comment (sanchitintel) [reply torch/_inductor/fx_passes/fuse_attention.py:479]: It did not match when `torch.finfo(dtype).min` was used. Should I add a comment there?

Comment (Valentine233) [reply torch/_inductor/fx_passes/fuse_attention.py:479]: Could you please elaborate on why `torch.finfo(dtype).min` cannot be matched? Is it the same issue as https://github.com/pytorch/pytorch/issues/117735?

Comment (sanchitintel) [reply torch/_inductor/fx_passes/fuse_attention.py]: Sorry, I had forgotten to add a `torch.full` node while I was testing. The pattern matched.

**Review by Valentine233 — APPROVED** [review]
LGTM

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 4]
```diff
@@ -704,6 +756,22 @@ def _get_sfdp_patterns():
                 d,
                 _sfdp_extra_check(aten.div.Tensor),
             ),
+            (
+                _sfdp_pattern_18,
+                _sfdp_replacement_18,
+                [g(), g(), g(), m_bool()],
+                d,
+                # CUDA AOT Inductor CI job's GPT2ForSequenceClassification accuracy test failed
+                _sfdp_extra_check(disable_cuda=True),
```
Comment (eellison) [inline torch/_inductor/fx_passes/fuse_attention.py:765]: Would be great if we could resolve this before landing. cc @desertfire @drisspg.

What was the error ?

Comment (sanchitintel) [reply torch/_inductor/fx_passes/fuse_attention.py:765]: Hi @eellison, thanks a lot for reviewing!

`GPT2ForSequenceClassification` had failed in the [AOT Inductor CI job for CUDA](https://github.com/pytorch/pytorch/actions/runs/8319111885/job/22762567770) because its accuracy check had failed - 

image

Comment (eellison) [reply torch/_inductor/fx_passes/fuse_attention.py:765]: It's strange that AOT Inductor fails but none of the other inductor jobs

Comment (desertfire) [reply torch/_inductor/fx_passes/fuse_attention.py:765]: It is strange indeed. Can you check the output code to see what's different?

Comment (eellison) [reply torch/_inductor/fx_passes/fuse_attention.py:765]: We could just skip the check if aot_inductor is enabled ? although it would be better to fix

Comment (sanchitintel) [reply torch/_inductor/fx_passes/fuse_attention.py:765]: Good point, thanks! It looks like those accuracy tests are only being run in that particular CI job, and are only being run for `--export-aot-inductor` -

https://github.com/pytorch/pytorch/blob/main/.ci/pytorch/test.sh#L472-L478

Comment (sanchitintel) [reply torch/_inductor/fx_passes/fuse_attention.py:765]: Sorry, the last two review comments just became visible at my end.

> Can you check the output code to see what's different?

@desertfire, sorry, I don't currently have access to an Nvidia GPU, so I'm currently unable to debug it.

> We could just skip the check if aot_inductor is enabled ?

@eellison, please clarify if you're referring to disabling `GPT2ForSequenceClassification` accuracy-check for `--export-aot-inductor` (by ignoring the error in `benchmarks/dynamo/ci_expected_accuracy/aot_inductor_huggingface_inference.csv`), or if you meant there's a way to skip replacement of that pattern in case of AOT Inductor export.

Thanks!

**Comment by sanchitintel** [comment]
Hi @eellison, the CI failure is unrelated, as it's a MacOS failure for `test/test_ops_fwd_gradients.py::TestFwdGradientsCPU::test_fn_fwgrad_bwgrad_nn_functional_conv_transpose3d_cpu_complex128`.

I re-enabled the pattern for CUDA & created a new GitHub issue to track `GPT2ForSequenceClassification`'s accuracy failure for CUDA AOT Inductor export CI job.

Thanks!

File: benchmarks/dynamo/ci_expected_accuracy/aot_inductor_huggingface_inference.csv [hunk 5]
```diff
@@ -82,7 +82,7 @@ ElectraForQuestionAnswering,pass,0
 
 
 
-GPT2ForSequenceClassification,pass,0
+GPT2ForSequenceClassification,fail_accuracy,0
```
Comment (jgong5) [inline benchmarks/dynamo/ci_expected_accuracy/aot_inductor_huggingface_inference.csv:85]: Can we temporarily disable the fusion for cuda before the issue is fixed? Otherwise, it is a regression.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: benchmarks/dynamo/ci_expected_accuracy/aot_inductor_huggingface_inference.csv (hunk 5 above)
Comment (sanchitintel) [reply benchmarks/dynamo/ci_expected_accuracy/aot_inductor_huggingface_inference.csv:85]: Hi @jgong5, please refer to [this comment](https://github.com/pytorch/pytorch/pull/121866#discussion_r1534373202) from @eellison. At the time Elias commented, CUDA support was already disabled for this fusion, so I think he's alluding to skip this particular test.
Moreover, for CUDA, only AOT Inductor export path is being affected.

Please advise if I should still disable this pattern-matching for CUDA. Thanks!

Comment (jgong5) [reply benchmarks/dynamo/ci_expected_accuracy/aot_inductor_huggingface_inference.csv:85]: But it is still a regression, right? I think we'd better avoid introducing regressions.

Comment (sanchitintel) [reply benchmarks/dynamo/ci_expected_accuracy/aot_inductor_huggingface_inference.csv]: Reverted to the penultimate commit. Thanks!

**Review by jgong5 — APPROVED** [review]

**Review by desertfire — APPROVED** [review]
